### PR TITLE
Improve diagnostics for constants being used in irrefutable patterns

### PR DESCRIPTION
**Pull request by Manishearth**
It's pretty confusing and this error triggers in resolve only when "shadowing" a const, so let's make that clearer.

r? @steveklabnik

### CODE REVIEW

File: src/librustc_resolve/diagnostics.rs [hunk 1]
```diff
@@ -623,6 +623,73 @@ let Foo = 12i32; // ok!
 The goal here is to avoid a conflict of names.
 "##,
 
+
```
Comment (GuillaumeGomez) [inline src/librustc_resolve/diagnostics.rs]: Extra empty line.

File: src/librustc_resolve/diagnostics.rs [hunk 2]
```diff
@@ -623,6 +623,73 @@ let Foo = 12i32; // ok!
...
+// or
+
+for FOO in bar {
+
+}
+```
+
+Introducing a new variable in Rust is done through a pattern. Thus you can have
+`let` bindings like `let (a,b) = ...`. However, patterns also allow constants in
```
Comment (GuillaumeGomez) [inline src/librustc_resolve/diagnostics.rs]: Small nit: "let (a, b) = ...".

File: src/librustc_resolve/diagnostics.rs [hunk 3]
```diff
@@ -623,6 +623,73 @@ let Foo = 12i32; // ok!
...
+them, e.g. if you want to match over a constant:
+
```
Comment (GuillaumeGomez) [inline src/librustc_resolve/diagnostics.rs]: Extra empty line.

File: src/librustc_resolve/diagnostics.rs [hunk 4]
```diff
@@ -623,6 +623,73 @@ let Foo = 12i32; // ok!
...
+let SOME_CONST = foo; // what if foo is not SOME_CONST?
+```
+
+Thus, an irrefutable variable binding can't contain a constant.
+
+To fix this error, just give the marked variable a different name.
+"##,
+
+
```
Comment (GuillaumeGomez) [inline src/librustc_resolve/diagnostics.rs]: Extra empty line.

**Comment by GuillaumeGomez** [comment]
Hum... Like you said, it's a bit confusing when you read it. The best explanation I could find for this is maybe too much details? Hard to say. :-/

**Comment by Manishearth** [comment]
updated

**Comment by GuillaumeGomez** [comment]
Great job! Thanks a lot @Manishearth!

r=me @steveklabnik

**Pull request merged**